nondex-common/src/main/java/edu/illinois/nondex/common/SystematicRandom.java

```diff
@@ -0,0 +1,144 @@
+a copy of this software and associated documentation files (the
+"Software"), to deal in the Software without restriction, including
+without limitation the rights to use, copy, modify, merge, publish,
+distribute, sublicense, and/or sell copies of the Software, and to
+permit persons to whom the Software is furnished to do so, subject to
+the following conditions:
+
+The above copyright notice and this permission notice shall be
+included in all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+package edu.illinois.nondex.common;
+
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.Random;
+import java.util.Stack;
+import java.util.logging.Level;
+
+public class SystematicRandom extends Random {
+    private final Stack<ExplorationEntry> choices;
+    private final String logFileName;
+    private final Configuration config = Configuration.parseArgs();
+    private List<String> lines;
+    private int replayIndex;
+
+    public SystematicRandom() {
+        logFileName = config.systematicLog;
+        choices = new Stack<ExplorationEntry>();
+        File file = new File(logFileName);
+        if (file.exists()) {
+            try {
+                lines = Files.readAllLines(Paths.get(logFileName));
+            } catch (IOException ioe) {
+                Logger.getGlobal().log(Level.SEVERE,"Could not read lines from systematic.log" ,ioe);
+            }
+            Object[] choiceValues;
+            for (String element: lines) {
+                String delimiter = "[ ]+";
+                if (!element.isEmpty()) {
+                    choiceValues = element.split(delimiter);
+                    if (choiceValues.length == 3) {
+                        int current = Integer.parseInt(choiceValues[0].toString());
+                        int maximum = Integer.parseInt(choiceValues[1].toString());
+                        boolean shouldExplore = Boolean.parseBoolean(choiceValues[2].toString());
+                        ExplorationEntry currentMaximum = new ExplorationEntry(current, maximum, shouldExplore);
+                        choices.push(currentMaximum);
+                    } else {
+                        Logger.getGlobal().log(Level.SEVERE, "The 3 ExplorationEntry variable were not stored properly");
+                    }
+
+                }
+            }
+        }
+    }
+
+    public int nextInt(final int maximum) {
```

Comment:
Add the following test:
x = nextInt(3);
if (x == 1)
  y = nextInt(2)
else
  y = nextInt(3)
print x,y

Your exploration should explore:
0,0
0,1
0,1
1,0
1,1
2,0
2,1
2,2

I believe you will explore 1,2 also, which is incorrect. Write a test and run it to confirm. Get back here if that's the case and we can think how to fix it. Otherwise I may be misunderstanding.